File: src/algorithms/graphs/depth-first-search/solution1.js
// STRATEGY:
// For every node, add that node's name to final array.
// For each child, call DFS and pass in child array.
// COMPLEXITY:
// Dependent on vertices (V) and edges (E)
// O(V + E) Time | O(V) Space

class Node {
    constructor(name) {
        this.name = name;
        this.children = [];
    }

    addChild(name) {
        this.children.push(new Node(name));
        return this;
    }

    depthFirstSearch(array) {
        array.push(this.name);
        for (const child of this.children) {
            child.depthFirstSearch(array);
        }
        return array;
    }
}

exports.Node = Node;

File: src/algorithms/graphs/depth-first-search/solution1.py
# STRATEGY:
# For every node, add that node's name to final array.
# For each child, call DFS and pass in child array.
# COMPLEXITY:
# Dependent on vertices (V) and edges (E)
# O(V + E) Time | O(V) Space

class Node:
    def __init__(self, name):
        self.children = []
        self.name = name
        
    def addChild(self, name):
        self.children.append(Node(name))
        return self

    def depthFirstSearch(self, array):
        array.append(self.name)
        for child in self.children:
            child.depthFirstSearch(array)
        return array